core/src/main/scala/kafka/zk/ZkData.scala

```diff
@@ -493,26 +495,54 @@ object ResourceZNode {
   def decode(bytes: Array[Byte], stat: Stat): VersionedAcls = VersionedAcls(Acl.fromBytes(bytes), stat.getVersion)
 }
 
+object AclChangeNotificationZNode {
+  def path = "/kafka-acl-changes"
+}
+
+case class AclChangeEvent(@BeanProperty @JsonProperty("version") version: Int,
+                          @BeanProperty @JsonProperty("resourceType") resourceType: String,
+                          @BeanProperty @JsonProperty("name") name: String,
+                          @BeanProperty @JsonProperty("resourceNameType") resourceNameType: String) {
+  if (version > AclChangeEvent.currentVersion)
+    throw new UnsupportedVersionException(s"Acl change event received for unsupported version: $version")
+
+  def toResource : Try[Resource] = {
+    for {
+      resType <- Try(ResourceType.fromString(resourceType))
+      nameType <- Try(ResourceNameType.fromString(resourceNameType))
+      resource = Resource(resType, name, nameType)
+    } yield resource
+  }
+}
+
+object AclChangeEvent {
+  val currentVersion: Int = 1
+}
+
 object AclChangeNotificationSequenceZNode {
-  val Separator = ":"
   def SequenceNumberPrefix = "acl_changes_"
 
-  def encode(resource: Resource): Array[Byte] = {
-    (resource.resourceType.name + Separator + resource.name).getBytes(UTF_8)
-  }
+  def createPath = s"${AclChangeNotificationZNode.path}/$SequenceNumberPrefix"
+  def deletePath(sequenceNode: String) = s"${AclChangeNotificationZNode.path}/$sequenceNode"
 
-  def decode(nameType: ResourceNameType, bytes: Array[Byte]): Resource = {
-    val str = new String(bytes, UTF_8)
-    str.split(Separator, 2) match {
-      case Array(resourceType, name, _*) => Resource(ResourceType.fromString(resourceType), name, nameType)
-      case _ => throw new IllegalArgumentException("expected a string in format ResourceType:ResourceName but got " + str)
+  def encode(resource: Resource): Array[Byte] =
+    Json.encodeAsBytes(AclChangeEvent(
+      AclChangeEvent.currentVersion,
+      resource.resourceType.name,
+      resource.name,
+      resource.resourceNameType.name))
+
+  def decode(bytes: Array[Byte]): Resource = {
+    val changeEvent = Json.parseBytesAs[AclChangeEvent](bytes) match {
+      case Right(event) => event
+      case Left(e) => throw new IllegalArgumentException("Failed to parse ACL change event", e)
```

Comment:
During the upgrade phase, the new broker may see an acl_change node of the old format. So, we want to fall back to parse the value of the node using the old format if it's not json.

Reply:
Done